src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java

```diff
@@ -0,0 +1,570 @@
+                + "failed_attempts INT UNSIGNED NOT NULL," + "PRIMARY KEY (device_id_hash));";
+    }
+
+    public static String getQueryToCreateCodesTable(Start start) {
+        return "CREATE TABLE IF NOT EXISTS " + Config.getConfig(start).getPasswordlessCodesTable() + " ("
+                + "code_id CHAR(36) NOT NULL," + "device_id_hash CHAR(44) NOT NULL,"
+                + "link_code_hash CHAR(44) NOT NULL UNIQUE," + "created_at BIGINT UNSIGNED NOT NULL,"
+                + "PRIMARY KEY (code_id),"
+                + "FOREIGN KEY (device_id_hash) REFERENCES passwordless_devices(device_id_hash) ON DELETE CASCADE ON UPDATE CASCADE);";
+    }
+
+    public static String getQueryToCreateDeviceEmailIndex(Start start) {
+        return "CREATE INDEX passwordless_devices_email_index ON "
+                + Config.getConfig(start).getPasswordlessDevicesTable() + " (email);"; // USING hash
+    }
+
+    public static String getQueryToCreateDevicePhoneNumberIndex(Start start) {
+        return "CREATE INDEX passwordless_devices_phone_number_index ON "
+                + Config.getConfig(start).getPasswordlessDevicesTable() + " (phone_number);"; // USING hash
+    }
+
+    public static String getQueryToCreateCodeCreatedAtIndex(Start start) {
+        return "CREATE INDEX passwordless_codes_created_at_index ON "
+                + Config.getConfig(start).getPasswordlessCodesTable() + "(created_at);";
+    }
+
+    public static void createDeviceWithCode(Start start, String email, String phoneNumber, PasswordlessCode code)
+            throws StorageTransactionLogicException, StorageQueryException {
+        start.startTransaction(con -> {
+            Connection sqlCon = (Connection) con.getConnection();
+            try {
+                String QUERY = "INSERT INTO " + Config.getConfig(start).getPasswordlessDevicesTable()
+                        + "(device_id_hash, email, phone_number, failed_attempts)" + " VALUES(?, ?, ?, 0)";
+                try (PreparedStatement pst = sqlCon.prepareStatement(QUERY)) {
+                    pst.setString(1, code.deviceIdHash);
+                    pst.setString(2, email);
+                    pst.setString(3, phoneNumber);
+                    pst.executeUpdate();
+                }
+
+                PasswordlessQueries.createCode_Transaction(start, sqlCon, code);
+                sqlCon.commit();
+            } catch (SQLException throwables) {
+                throw new StorageTransactionLogicException(throwables);
+            }
+            return null;
+        });
+    }
+
+    public static PasswordlessDevice getDevice_Transaction(Start start, Connection con, String deviceIdHash)
+            throws StorageQueryException, SQLException {
+
+        String QUERY = "SELECT device_id_hash, email, phone_number, failed_attempts FROM "
+                + Config.getConfig(start).getPasswordlessDevicesTable() + " WHERE device_id_hash = ?";
```

Comment:
We need to lock the device using FOR UPDATE

Reply:
@rishabhpoddar @porcellus added : "LOCK IN SHARE MODE" instead of "FOR UPDATE". Does that sound good ?

Reply:
No, we want FOR UPDATE here.